store_bulk_ops.go

```diff
@@ -0,0 +1,686 @@
+		backgroundFetchWorkers,
+		humanize.Comma(int64(len(ordinalsUnique))),
+		float64(time.Since(t0).Milliseconds())/1000,
+	)
+
+	ordinalList := make([]int64, 0, len(ordinalsUnique))
+	for o := range ordinalsUnique {
+		ordinalList = append(ordinalList, o)
+	}
+	ordinalsUnique = nil // release early/explicitly
+
+	// might as well sort (descending) - potential sequential page reads
+	sort.Slice(ordinalList, func(i, j int) bool {
+		return ordinalList[j] < ordinalList[i]
+	})
+
+	t0 = time.Now()
+	totalCount, totalBytes, err := dbbs.PrecacheByOrdinal(ctx, ordinalList)
+
+	tookSecs := float64(time.Since(t0).Milliseconds()) / 1000
+
+	if err != nil {
+		log.Warnf(
+			"priming of the write-through LRU cache on startup ABORTED after %s blocks totalling %s bytes after %.03fs at %.02fMiB/s: %s",
+			humanize.Comma(totalCount), humanize.Comma(totalBytes),
+			tookSecs, float64(totalBytes/(1<<20))/tookSecs,
+			err,
+		)
+	} else {
+		log.Infof(
+			"successfully primed the write-through LRU cache on startup with %s recently-accessed blocks totalling %s bytes: took %.03fs at %.02fMiB/s",
+			humanize.Comma(totalCount), humanize.Comma(totalBytes),
+			tookSecs, float64(totalBytes/(1<<20))/tookSecs,
+		)
+	}
+}
+
+// PrecacheByOrdinal primes the cache with a given (potentially very large)
+// list of block ordinals. At scale (millions of blocks) it is not practical to
+// read blocks sequentially, both due to performance and also due to the RDBMS
+// tending to MVCC-lock the entire set until we go through it. Instead this
+// method breaks down the list into batches and performs parallel relatively
+// short reads until everything is retrieved.
+func (dbbs *PgBlockstore) PrecacheByOrdinal(ctx context.Context, ordinals []int64) (totalCount, totalBytes int64, err error) {
+
+	var wg sync.WaitGroup
+	totCount := new(int64)
+	totBytes := new(int64)
+
+	limiter := make(chan struct{}, backgroundFetchWorkers)
+	errs := make(chan error, backgroundFetchWorkers)
+	for len(ordinals) > 0 && len(errs) == 0 {
+
+		// we have exceeded our cache range - no point continuing
+		if atomic.LoadInt64(totBytes) > dbbs.lruSizeBytes {
+			break
+		}
+
+		limiter <- struct{}{}
+		wg.Add(1)
+
+		batchLen := len(ordinals)
+		if batchLen > BulkFetchSize {
+			batchLen = BulkFetchSize
+		}
+
+		batch := ordinals[len(ordinals)-batchLen:]
+		ordinals = ordinals[:len(ordinals)-batchLen]
+
+		go func() {
+			defer func() {
+				<-limiter
+				wg.Done()
+			}()
+
+			rows, err := dbbs.dbPool.Query(
+				ctx,
+				fmt.Sprintf(
+					`
+					SELECT %s
+						FROM fil_common_base.datablocks
+					WHERE
+						block_ordinal = ANY( $1::BIGINT[] )
+							AND
+						content IS NOT NULL
+					`,
+					StoredBlocksInflatorSelection,
+				),
+				batch,
+			)
+
+			if err != nil {
+				if rows != nil {
+					rows.Close()
+				}
+				errs <- err
+				return
+			}
+
+			sbs, bytes, err := dbbs.InflateDbRows(rows, false, false)
+			if err != nil {
+				errs <- err
+				return
```

Comment:
@iand no because the inflator does that. I renamed the method to make it more sensible: https://github.com/filecoin-project/go-bs-postgres-chainnotated/pull/1/commits/84c363ecaa6